domain/machine/modelmigration/export.go

```diff
@@ -22,34 +24,97 @@ func RegisterExport(coordinator Coordinator, logger logger.Logger) {
 }
 
 // exportOperation describes a way to execute a migration for
-// exporting machines.
+// exporting external controllers.
 type exportOperation struct {
 	modelmigration.BaseOperation
 
-	logger  logger.Logger
 	service ExportService
+	logger  logger.Logger
 }
 
 // Name returns the name of this operation.
 func (e *exportOperation) Name() string {
 	return "export machines"
 }
 
-// Setup implements Operation.
 func (e *exportOperation) Setup(scope modelmigration.Scope) error {
-	e.service = service.NewService(
-		state.NewState(scope.ModelDB(), e.logger))
+	e.service = service.NewService(state.NewState(scope.ModelDB(), e.logger))
 	return nil
 }
 
-// Execute the export, adding the machine to the model.
 func (e *exportOperation) Execute(ctx context.Context, model description.Model) error {
+
+	for _, machine := range model.Machines() {
+		// TODO(nvinuesa): We must check if the machine cloud instance is already
+		// set and in that case don't ovewrite anything. This check can be removed
+		// once we fully move machines over to dqlite.
+		if instance := machine.Instance(); instance != nil {
+			continue
+		}
+
+		machineName := coremachine.Name(machine.Id())
+		machineUUID, err := e.service.GetMachineUUID(ctx, machineName)
+		if err != nil {
+			return errors.Errorf("retrieving instance ID for machine %q: %w", machineName, err)
+		}
+		instanceID, err := e.service.InstanceID(ctx, machineUUID)
+		if err != nil {
+			return errors.Errorf("retrieving instance ID for machine %q: %w", machineName, err)
+		}
+		instanceArgs := description.CloudInstanceArgs{
+			InstanceId: instanceID,
+		}
+		hardwareCharacteristics, err := e.service.HardwareCharacteristics(ctx, machineUUID)
+		if err != nil {
+			return errors.Errorf("retrieving hardware characteristics for machine %q: %w", machineName, err)
+		}
+		if hardwareCharacteristics.Arch != nil {
+			instanceArgs.Architecture = *hardwareCharacteristics.Arch
+		}
+		if hardwareCharacteristics.Mem != nil {
+			instanceArgs.Memory = *hardwareCharacteristics.Mem
+		}
+		if hardwareCharacteristics.RootDisk != nil {
+			instanceArgs.RootDisk = *hardwareCharacteristics.RootDisk
+		}
+		if hardwareCharacteristics.RootDiskSource != nil {
+			instanceArgs.RootDiskSource = *hardwareCharacteristics.RootDiskSource
+		}
+		if hardwareCharacteristics.CpuCores != nil {
+			instanceArgs.CpuCores = *hardwareCharacteristics.CpuCores
+		}
+		if hardwareCharacteristics.CpuPower != nil {
+			instanceArgs.CpuPower = *hardwareCharacteristics.CpuPower
+		}
+		if hardwareCharacteristics.Tags != nil {
+			instanceArgs.Tags = *hardwareCharacteristics.Tags
+		}
+		if hardwareCharacteristics.AvailabilityZone != nil {
+			instanceArgs.AvailabilityZone = *hardwareCharacteristics.AvailabilityZone
+		}
+		if hardwareCharacteristics.VirtType != nil {
+			instanceArgs.VirtType = *hardwareCharacteristics.VirtType
+		}
+		machine.SetInstance(instanceArgs)
+		instance := machine.Instance()
+		instance.SetStatus(description.StatusArgs{})
```

Comment:
We require this because it fails if memory serves. Is that correct?

Reply:
Indeed, we cannot serialize description model without it.